pkg/reconcilers/authorino_reconciler.go

```diff
@@ -130,131 +139,173 @@ func (r *AuthorinoReconciler) ReconcileAuthorinoServices(ctx context.Context, lo
-	} else {
-		roleBinding = authorinoResources.GetAuthorinoRoleBinding(authorino.Namespace, authorino.Name, roleBindingName, "ClusterRole", clusterRoleName, serviceAccount, authorino.Labels)
-		roleBinding.SetNamespace(authorino.Namespace)
+func (r *AuthorinoReconciler) reconcileManagerAuthClusterRoleBinding(ctx context.Context, authorinoInstance *api.Authorino) error {
+	clusterRoleKey := client.ObjectKey{Name: AuthorinoK8sAuthClusterRoleName}
+	if err := r.checkClusterRoleExists(ctx, clusterRoleKey, authorinoInstance); err != nil {
+		return err
 	}
 
-	roleBinding = authorinoResources.AppendSubjectToRoleBinding(roleBinding, serviceAccount)
+	sa := authorinoResources.GetAuthorinoServiceAccount(authorinoInstance.Namespace, authorinoInstance.Name, authorinoInstance.Labels)
 
-	return roleBinding, nil
+	binding := authorinoResources.GetAuthorinoClusterRoleBinding(AuthorinoK8sAuthClusterRoleBindingName, AuthorinoK8sAuthClusterRoleName, sa)
+	return r.reconcileClusterRoleBinding(
+		ctx, binding,
+		ClusterRoleBindingMutator(ClusterRoleBindingLabelsMutator, ClusterRoleBindingSubjectMutator),
+		authorinoInstance,
+	)
 }
 
-func (r *AuthorinoReconciler) bindAuthorinoServiceAccountToLeaderElectionRole(authorino *api.Authorino, serviceAccount k8score.ServiceAccount) error {
-	var logger = r.Log
+func (r *AuthorinoReconciler) reconcileLeaderElectionRole(ctx context.Context, authorinoInstance *api.Authorino) error {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return err
+	}
 
-	leaderElectionRole := &k8srbac.Role{}
-	leaderElectionNsdName := namespacedName(authorino.Namespace, AuthorinoLeaderElectionRoleName)
-	if err := r.Get(context.TODO(), leaderElectionNsdName, leaderElectionRole); err != nil {
-		if errors.IsNotFound(err) {
-			// leader election Role doesn't exist then create
-			leaderElectionRole.Name = AuthorinoLeaderElectionRoleName
-			leaderElectionRole.Namespace = authorino.Namespace
-			leaderElectionRole.Rules = authorinoResources.GetLeaderElectionRules()
-			_ = ctrl.SetControllerReference(authorino, leaderElectionRole, r.Scheme)
-			if err := r.Client.Create(context.TODO(), leaderElectionRole); err != nil {
-				return r.WrapErrorWithStatusUpdate(
-					logger, authorino, r.SetStatusFailed(statusUnableToCreateLeaderElectionRole),
-					fmt.Errorf("failed to create %s role, err: %v", leaderElectionRole, err),
-				)
-			}
+	role := &k8srbac.Role{
+		TypeMeta:   k8smeta.TypeMeta{APIVersion: k8srbac.SchemeGroupVersion.Version, Kind: "Role"},
+		ObjectMeta: k8smeta.ObjectMeta{Name: AuthorinoLeaderElectionRoleName, Namespace: authorinoInstance.Namespace},
+		Rules:      authorinoResources.GetLeaderElectionRules(),
+	}
+
+	if err := ctrl.SetControllerReference(authorinoInstance, role, r.Scheme); err != nil {
+		return err
+	}
+
+	crud, _, err := r.reconcileResource(ctx, &k8srbac.Role{}, role, CreateOnlyMutator)
+	if err != nil {
+		if crud == "read" {
```

Comment:
Something in how we're controlling the flow based on a string returned by the method smells.

Ideally, we could leverage proper typing here instead, by returning an interface that abstracts the details of building the error according to each operation.

I understand this would involve a bigger change, so maybe a couple alternatives to mitigate and reduce the risks could be:
- returning the error already wrapped
- using a constant to identify the operation

Reply:
I was thinking about that as well. I agree it would be much better than string. 

Error type?

I was also looking at the object being returned by the `reconcileResource`.. it might not be necessary. But for follow up work